sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -366,14 +366,25 @@ class CacheManager extends Logging with AdaptiveSparkPlanHelper {
       fileIndex: FileIndex,
       fs: FileSystem,
       qualifiedPath: Path): Boolean = {
-    val prefixToInvalidate = qualifiedPath.toString
     val needToRefresh = fileIndex.rootPaths
-      .map(_.makeQualified(fs.getUri, fs.getWorkingDirectory).toString)
-      .exists(_.startsWith(prefixToInvalidate))
+      .map(_.makeQualified(fs.getUri, fs.getWorkingDirectory))
+      .exists(isSubDir(qualifiedPath, _))
     if (needToRefresh) fileIndex.refresh()
     needToRefresh
   }
 
+  /**
+   * Checks if the given child path is a sub-directory of the given parent path.
+   * @param qualifiedPathChild: Fully qualified child path
+   * @param qualifiedPathParent: Fully qualified parent path.
+   * @return True if the child path is a sub-directory of the given parent path. Otherwise, false.
+   */
+  def isSubDir(qualifiedPathParent: Path, qualifiedPathChild: Path): Boolean = {
+    Iterator.iterate(qualifiedPathChild)(_.getParent)
+      .takeWhile(_ != null)
+      .exists(_.toString.equals(qualifiedPathParent.toString))
```

Comment:
can't we rely on `Path.equals`?

Reply:
I think we can. Let me change this. to use `Path.equals`